tools/deploy.go

```diff
@@ -0,0 +1,711 @@
+			continue
+		}
+
+		obj := &unstructured.Unstructured{}
+		_, gvk, err := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme).Decode(doc, nil, obj)
+		if err != nil {
+			return fmt.Errorf("error decoding YAML document: %w", err)
+		}
+
+		mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err != nil {
+			return fmt.Errorf("error finding mapping for GVK %s: %w", gvk.String(), err)
+		}
+
+		resourceClient := dynClient.Resource(mapping.Resource).Namespace(obj.GetNamespace())
+		err = applyResource(resourceClient, obj)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// applyResource creates or updates a given resource.
+func applyResource(resourceClient dynamic.ResourceInterface, obj *unstructured.Unstructured) error {
+	ctx := context.Background()
+	name := obj.GetName()
+
+	existingObj, err := resourceClient.Get(ctx, name, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		fmt.Printf("Creating %s %q\n", obj.GetKind(), name)
+		_, err := resourceClient.Create(ctx, obj, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("error creating %s %q: %w", obj.GetKind(), name, err)
+		}
+	} else if err == nil {
+		fmt.Printf("Updating %s %q\n", obj.GetKind(), name)
+		obj.SetResourceVersion(existingObj.GetResourceVersion())
+		_, err := resourceClient.Update(ctx, obj, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("error updating %s %q: %w", obj.GetKind(), name, err)
+		}
+	} else {
+		return fmt.Errorf("error getting %s %q: %w", obj.GetKind(), name, err)
+	}
+
+	return nil
+}
+
+func execCommandToFile(cmd *exec.Cmd, outputPath string) error {
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("command execution failed: %w", err)
+	}
+
+	err = os.WriteFile(outputPath, output, 0600)
+	if err != nil {
+		return fmt.Errorf("writing command output to file failed: %w", err)
+	}
+
+	return nil
+}
+
+// deployWithKustomizeAndApply first generates the YAML configuration by running Kustomize build on the overlay directory,
+// then outputs the generated YAML to a temp file within the overlay dir, and then applies
+// the configuration to the Kubernetes cluster specified in the kubeconfig file.
+func deployWithKustomizeAndApply(kustomizePath, overlayPath string) error {
+	yamlOutputPath := filepath.Join(overlayPath, "output.yaml")
+
+	defer func() {
+		if removeErr := os.Remove(yamlOutputPath); removeErr != nil {
+			log.Printf("Warning: Failed to remove temporary YAML file %s: %v", yamlOutputPath, removeErr)
+		}
+	}()
+
+	// Generate YAML with kustomize
+	cmd := exec.Command(kustomizePath, "build", overlayPath)
+	err := execCommandToFile(cmd, yamlOutputPath)
+	if err != nil {
+		return fmt.Errorf("failed to generate YAML with kustomize: %v", err)
+	}
+
+	err = applyYAML(yamlOutputPath)
+	if err != nil {
+		return fmt.Errorf("failed to apply YAML: %v", err)
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string) error {
```

Comment:
What is different from `os.Copy()`?

Reply:
I don't still see why you're not using `os.Copy()`?

Reply:
I need file descriptors for `io.Copy()`. I now use io.Copy() in my `copyFile` function.